encrypted_json_fields/encryption.py

```diff
@@ -0,0 +1,305 @@
+from abc import ABC, abstractmethod
+import json
+from cryptography.fernet import Fernet, MultiFernet, InvalidToken
+from typing import Union
+import base64
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, hmac
+from cryptography.exceptions import InvalidSignature
+import os
+from ast import literal_eval
+from .utils import pad, unpad
+from .constants import AES_PREFIX, FERNET_PREFIX
+
+class MultiAES:
+    def __init__(self, keys):
+        for key in keys:
+            if len(key) != 32:
+                raise ValueError("Invalid AES key length: must be 32 bytes.")
+        self.keys = keys
+
+    def encrypt(self, plaintext: bytes) -> bytes:
+        key = self.keys[-1]
+        if len(key) != 32:
+            raise ValueError("AES key must be 32 bytes.")
+        block_size = 16
+        iv = os.urandom(block_size)
+        cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
+        encryptor = cipher.encryptor()
+
+        padded_data = pad(plaintext, block_size)
+        ciphertext = encryptor.update(padded_data) + encryptor.finalize()
+
+        h = hmac.HMAC(key, hashes.SHA256())
+        h.update(iv + ciphertext)
+        hmac_tag = h.finalize()
+
+        return iv + ciphertext + hmac_tag
+
+    def decrypt(self, ciphertext: bytes) -> bytes:
+        block_size = 16
+        for key in self.keys:
+            try:
+                if len(key) != 32:
+                    raise ValueError("AES key must be 32 bytes.")
+                iv = ciphertext[:block_size]
+                cipher_text_only = ciphertext[block_size:-32]
+                hmac_tag = ciphertext[-32:]
+
+                h = hmac.HMAC(key, hashes.SHA256())
+                h.update(iv + cipher_text_only)
+                h.verify(hmac_tag)
+
+                cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
+                decryptor = cipher.decryptor()
+                padded_data = decryptor.update(cipher_text_only) + decryptor.finalize()
+                return unpad(padded_data, block_size)
+            except (InvalidSignature, ValueError):
+                continue
+        raise ValueError("Decryption failed with all keys")
+
+class EncryptionMethod(ABC):
```

Comment:
coming back to this it's a little hard to keep track of what each method is doing.

internal vs raw vs legacy, etc.. Can we add some good docstrings to each of these and help explain how/what/when/why each of these are for.